### ISSUE TITLE
#error machine_vector.h not implemented on AMD Ryzen 7 without forcing --enable-avx

### ISSUE THREAD

**Issue by jerebertho** [flintlib/flint#2031]
Dear all,

On the current commit of Flint (449694569163502fc31d66ac5363724969c9e756), compiling on my laptop with an AMD Ryzen 7 (see below for the output of cat /proc/cpuinfo) yields
```
In file included from ./src/fft_small.h:16,
                 from src/mpn_extras/mul.c:41:
./src/machine_vectors.h:1631:2: error: #error machine_vector.h not implemented
 1631 | #error machine_vector.h not implemented
      |  ^~~~~
  CC  mpn_extras/mulmod_2expp1_basecase.c
In file included from src/mpn_extras/mul.c:41:
./src/fft_small.h:329:11: error: unknown type name ‘vec4d’
  329 |     const vec4d* two_pow,
      |           ^~~~~
./src/fft_small.h:361:5: error: unknown type name ‘vec4d’
  361 |     vec4d* vec_two_pow_tab[(MPN_CTX_NCRTS + VEC_SZ - 1)/VEC_SZ];
      |     ^~~~~
./src/fft_small.h:362:5: error: unknown type name ‘vec4d’
  362 |     vec4d* vec_two_pow_buffer;
      |     ^~~~~
```

I need to set configure with --enable-avx2 and --enable-avx512 to make the problem disappear, yet I do not have to do these --enable-avx2 --enable-avx512 on my other computer with an Intel Xeon CPU.

My CPU is

> processor	: 15
> vendor_id	: AuthenticAMD
> cpu family	: 25
> model		: 116
> model name	: AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics
> stepping	: 1
> microcode	: 0xa704104
> cpu MHz		: 400.000
> cache size	: 1024 KB
> physical id	: 0
> siblings	: 16
> core id		: 7
> cpu cores	: 8
> apicid		: 15
> initial apicid	: 15
> fpu		: yes
> fpu_exception	: yes
> cpuid level	: 16
> wp		: yes
> flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d
> bugs		: sysret_ss_attrs spectre_v1 spectre_v2 spec_store_bypass srso
> bogomips	: 6587.66
> TLB size	: 3584 4K pages
> clflush size	: 64
> cache_alignment	: 64
> address sizes	: 48 bits physical, 48 bits virtual
> power management: ts ttp tm hwpstate cpb eff_freq_ro [13] [14] [15]
> 

Many thanks!

**Comment by albinahlback** [comment]
Could you post the output of running `configure`?

**Comment by jerebertho** [comment]
Sure, here it is!
```
checking build system type... zen4-pc-linux-gnu
checking host system type... zen4-pc-linux-gnu
checking how to print strings... printf
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert zen4-pc-linux-gnu file names to zen4-pc-linux-gnu format... func_convert_file_noop
checking how to convert zen4-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking for gawk... gawk
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... no
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for inline... inline
checking whether byte ordering is bigendian... no
checking if memory is strongly-ordered... yes
checking how to run the C preprocessor... gcc -E
checking for gmp.h... yes
checking if version of GMP is greater than 6.2.1... yes
checking if GMP defines mp_limb_t as unsigned long long int... no
checking for mpfr.h... yes
checking if version of MPFR is greater than 4.1.0... yes
checking for desired ABI... 64
checking for stdarg.h... yes
checking for math.h... yes
checking for float.h... yes
checking for errno.h... yes
checking for fenv.h... yes
checking for alloca.h... yes
checking for malloc.h... yes
checking for sys/param.h... yes
checking for windows.h... no
checking for pthread_np.h... no
checking for pthread.h... yes
checking for library containing atan2... -lm
checking for library containing __gmpz_init... -lgmp
checking for library containing __gmpn_mul_basecase... none required
checking for library containing __gmpn_gcd_11... none required
checking for library containing __gmpn_div_q... none required
checking for library containing __gmpn_invert_limb... none required
checking for library containing __gmpn_modexact_1_odd... none required
checking for library containing __gmpn_addmul_2... none required
checking for library containing __gmpn_add_n_sub_n... none required
checking for library containing __gmpn_add_nc... none required
checking for library containing __gmpn_sub_nc... none required
checking for library containing __gmpn_addlsh1_n... none required
checking for library containing __gmpn_addlsh1_n_ip1... no
checking for library containing __gmpn_rsh1sub_n... none required
checking for library containing __gmpn_rsh1add_n... none required
checking for library containing mpfr_init... -lmpfr
checking for library containing mpfr_round_p... none required
checking for library containing mpfr_mulhigh_n... none required
checking for library containing mpfr_sqrhigh_n... none required
checking whether gcc accepts -pthread... yes
checking if cpu_set_t is supported... yes
checking if alloca works... yes
checking for aligned_alloc... yes
checking for _aligned_malloc... no
checking if compiler is Clang... no
checking whether gcc accepts -O3... yes
checking whether gcc accepts -pedantic... yes
checking whether gcc accepts -std=c11... yes
checking whether gcc accepts -Werror=implicit-function-declaration... yes
checking whether gcc accepts -Wall... yes
checking whether gcc accepts -Wno-stringop-overread... yes
checking whether gcc accepts -Wno-stringop-overflow... yes
checking whether gcc accepts -Wno-maybe-uninitialized... yes
checking whether gcc accepts -march=znver4... no
checking whether gcc accepts -funroll-loops... yes
checking for immintrin.h... yes
checking if system have required x86_64 instruction set for fft_small... (cached) yes
checking if system can use FLINT's fft_small module... yes
checking for suitable m4... m4
checking how to switch to text section... .text
checking how to switch to data section... .data
checking for assembler label suffix... :
checking for assembler global directive... .globl
checking for assembler global directive attribute... 
checking if globals are prefixed by underscore... no
checking how to switch to read-only data section... 	.section	.rodata
checking for assembler .type directive... .type	$1,@$2
checking for assembler .size directive... .size	$1,$2
checking for assembler local label prefix... .L
checking for assembler byte directive... .byte
checking if .align assembly directive is logarithmic... no
checking if the .align directive accepts an 0x90 fill in .text... yes
checking for assembler COFF type directives... no
creating config.m4
configure: creating ./config.status
config.status: creating Makefile
config.status: creating flint.pc
config.status: creating src/flint.h
config.status: creating src/config.h
config.status: src/config.h is unchanged
config.status: creating src/flint-config.h
config.status: src/flint-config.h is unchanged
config.status: linking src/gmpcompat.h.in to src/gmpcompat.h
config.status: linking src/mpn_extras/x86_64/zen3/flint-mparam.h to src/flint-mparam.h
config.status: linking src/fmpz/link/fmpz_single.c to src/fmpz/fmpz.c
config.status: executing libtool commands
```

**Comment by jerebertho** [comment]
However if I configure with --enable-avx2 and --enable-avx512, here is the output
```
checking build system type... zen4-pc-linux-gnu
checking host system type... zen4-pc-linux-gnu
checking how to print strings... printf
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert zen4-pc-linux-gnu file names to zen4-pc-linux-gnu format... func_convert_file_noop
checking how to convert zen4-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking for gawk... gawk
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... no
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for inline... inline
checking whether byte ordering is bigendian... no
checking if memory is strongly-ordered... yes
checking how to run the C preprocessor... gcc -E
checking for gmp.h... yes
checking if version of GMP is greater than 6.2.1... yes
checking if GMP defines mp_limb_t as unsigned long long int... no
checking for mpfr.h... yes
checking if version of MPFR is greater than 4.1.0... yes
checking for desired ABI... 64
checking for stdarg.h... yes
checking for math.h... yes
checking for float.h... yes
checking for errno.h... yes
checking for fenv.h... yes
checking for alloca.h... yes
checking for malloc.h... yes
checking for sys/param.h... yes
checking for windows.h... no
checking for pthread_np.h... no
checking for pthread.h... yes
checking for library containing atan2... -lm
checking for library containing __gmpz_init... -lgmp
checking for library containing __gmpn_mul_basecase... none required
checking for library containing __gmpn_gcd_11... none required
checking for library containing __gmpn_div_q... none required
checking for library containing __gmpn_invert_limb... none required
checking for library containing __gmpn_modexact_1_odd... none required
checking for library containing __gmpn_addmul_2... none required
checking for library containing __gmpn_add_n_sub_n... none required
checking for library containing __gmpn_add_nc... none required
checking for library containing __gmpn_sub_nc... none required
checking for library containing __gmpn_addlsh1_n... none required
checking for library containing __gmpn_addlsh1_n_ip1... no
checking for library containing __gmpn_rsh1sub_n... none required
checking for library containing __gmpn_rsh1add_n... none required
checking for library containing mpfr_init... -lmpfr
checking for library containing mpfr_round_p... none required
checking for library containing mpfr_mulhigh_n... none required
checking for library containing mpfr_sqrhigh_n... none required
checking whether gcc accepts -pthread... yes
checking if cpu_set_t is supported... yes
checking if alloca works... yes
checking for aligned_alloc... yes
checking for _aligned_malloc... no
checking if compiler is Clang... no
checking whether gcc accepts -mavx2... yes
checking whether gcc accepts -mfma... yes
checking whether gcc accepts -mavx512f... yes
checking whether gcc accepts -O3... yes
checking whether gcc accepts -pedantic... yes
checking whether gcc accepts -std=c11... yes
checking whether gcc accepts -Werror=implicit-function-declaration... yes
checking whether gcc accepts -Wall... yes
checking whether gcc accepts -Wno-stringop-overread... yes
checking whether gcc accepts -Wno-stringop-overflow... yes
checking whether gcc accepts -Wno-maybe-uninitialized... yes
checking whether gcc accepts -march=znver4... no
checking whether gcc accepts -funroll-loops... yes
checking for immintrin.h... yes
checking if system have required x86_64 instruction set for fft_small... (cached) yes
checking if system can use FLINT's fft_small module... yes
checking for suitable m4... m4
checking how to switch to text section... .text
checking how to switch to data section... .data
checking for assembler label suffix... :
checking for assembler global directive... .globl
checking for assembler global directive attribute... 
checking if globals are prefixed by underscore... no
checking how to switch to read-only data section... 	.section	.rodata
checking for assembler .type directive... .type	$1,@$2
checking for assembler .size directive... .size	$1,$2
checking for assembler local label prefix... .L
checking for assembler byte directive... .byte
checking if .align assembly directive is logarithmic... no
checking if the .align directive accepts an 0x90 fill in .text... yes
checking for assembler COFF type directives... no
creating config.m4
configure: creating ./config.status
config.status: creating Makefile
config.status: creating flint.pc
config.status: creating src/flint.h
config.status: creating src/config.h
config.status: src/config.h is unchanged
config.status: creating src/flint-config.h
config.status: src/flint-config.h is unchanged
config.status: linking src/gmpcompat.h.in to src/gmpcompat.h
config.status: linking src/mpn_extras/x86_64/zen3/flint-mparam.h to src/flint-mparam.h
config.status: linking src/fmpz/link/fmpz_single.c to src/fmpz/fmpz.c
config.status: executing libtool commands
```
Where we can see at least these 3 extra lines
```
checking whether gcc accepts -mavx2... yes
checking whether gcc accepts -mfma... yes
checking whether gcc accepts -mavx512f... yes
```

**Comment by albinahlback** [comment]
Yeah, but that shouldn't matter. It should check which CPU you have and push the right flags accordingly.

**Comment by albinahlback** [comment]
Could you post the CFLAGS you get in the `Makefile` here?

**Comment by albinahlback** [comment]
And could you post your output of
```bash
echo | gcc <insert CFLAGS here> -dM -E - | grep "AVX"
```
here?

**Comment by jerebertho** [comment]
This?
```
CFLAGS:=-Wno-maybe-uninitialized -Wno-stringop-overflow -Wno-stringop-overread -Wall -Werror=implicit-function-declaration -std=c11 -pedantic -O3 -g 
```

**Comment by albinahlback** [comment]
> This?
> 
> ```
> CFLAGS:=-Wno-maybe-uninitialized -Wno-stringop-overflow -Wno-stringop-overread -Wall -Werror=implicit-function-declaration -std=c11 -pedantic -O3 -g 
> ```

Yeah, thanks! Looks like the `-march=...` is missing. Will try to see what is going on here.

**Comment by jerebertho** [comment]
> And could you post your output of
> 
> ```shell
> echo | gcc <insert CFLAGS here> -dM -E - | grep "AVX"
> ```
> 
> here?

I get nothing as the output.

**Comment by albinahlback** [comment]
Ah, I see. What version of GCC are you using? It doesn't seem to accept `-march=znver4`, which is the native option for Zen 4.

**Comment by jerebertho** [comment]
gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)

**Comment by albinahlback** [comment]
Alright. What I will do is push `-march=znver3` as a back-up option.

**Comment by jerebertho** [comment]
Thanks!

**Comment by albinahlback** [comment]
By the way, thanks for reporting this!

Solved via #2032
